torch/ao/quantization/dynamic_tracing/quantization_state.py

```diff
@@ -0,0 +1,725 @@
+
+    def validate_is_at_last_seen_idx(self) -> None:
+        is_at_last_seen_idx = (
+            len(self.idx_to_seen_ops) == 0 or
+            self.idx == len(self.idx_to_seen_ops)
+        )
+        assert is_at_last_seen_idx, \
+            f"Cur idx: {self.idx}, expected idx: {len(self.idx_to_seen_ops)}"
+
+    def extra_repr(self) -> str:
+        s = ""
+        # idx_to_seen_ops
+        if len(self.idx_to_seen_ops):
+            s += "(seen_ops): {\n"
+            for k, v in self.idx_to_seen_ops.items():
+                s += f"  {k}: {v}\n"
+            s += "}\n"
+        else:
+            s += "(seen_ops): {}\n"
+        # output_qtensor_infos
+        s += "(output_qtensor_infos): ["
+        for i in self.output_qtensor_infos:
+            s += f"{i} "
+        s += "]\n"
+        # idx_to_packed_weight_name
+        if len(self.idx_to_packed_weight_name):
+            s += "(idx_to_packed_weight_name): {\n"
+            for k, v in self.idx_to_packed_weight_name.items():  # type: ignore[assignment]
+                s += f"  {k}: {v}\n"
+            s += "}"
+        else:
+            s += "(idx_to_packed_weight_name): {}"
+        return s
+
+    def _get_cur_seen_op(self):
+        return self.idx_to_seen_ops[str(self.idx)]
+
+    def get_cur_output_inf_dtype(self):
+        return self._get_cur_seen_op().output_tensor_infos[0].inf_dtype
+
+    def reset_to_new_call(self):
+        """
+        Resets the internal op counter to start a new top level module call
+        """
+        self.idx = 0
+
+    def cur_op_needs_hooks(self, cur_op: Callable) -> bool:
+        return op_needs_quantization(cur_op)
+
+    def validate_cur_op(self, cur_op: Callable) -> None:
+        """
+        This function is expected to be called before any new function or
+        module call which needs hooks. It validates that the new function or
+        module is of the expected type based on the order of execution.
+        """
+        assert self.cur_op_needs_hooks(cur_op)
+        try:
+            seen_op = self._get_cur_seen_op()
+            expected_op = seen_op.type
+        except IndexError:
+            _raise_obs_not_found_error(cur_op)
+        if not ops_are_related(cur_op, expected_op):
+            _raise_obs_op_mismatch(cur_op, expected_op)
+
+    def mark_cur_op_complete(self, cur_op: Callable) -> None:
+        """
+        This function is expected to be called after a function or module
+        processing is complete.
+        """
+        if op_needs_quantization(cur_op):
+            self.idx += 1
+
+    def _outputs_prepare_hook_tensor(
+        self,
+        output: torch.Tensor,
+        first_call: bool,
+        qtensor_id: List[int],
+    ) -> torch.Tensor:
+        if first_call:
+            if not hasattr(output, '_qtensor_info'):
+                # TODO: use actual dtype instead of defaulting to float
+                output._qtensor_info = QTensorInfo(qtensor_id[0], torch.float)  # type: ignore[attr-defined]
```

Comment:
optional nit: should we name this to TensorInfo since it can be torch.float as well

Reply:
I'm open to better names.  `_tensor_info` sounds pretty generic, unfortunately, maybe we can come up with a name which is specific to quantization but doesn't look like it's only for quantized tensors.